src/Bicep.Core/Semantics/Namespaces/SystemNamespaceType.cs

```diff
@@ -1439,24 +1439,46 @@ static void ValidateLength(string decoratorName, DecoratorSyntax decoratorSyntax
                 })
                 .Build();
 
-            yield return new DecoratorBuilder(LanguageConstants.ExportPropertyName)
-                .WithDescription("Allows the type to be imported into other templates.")
-                .WithFlags(FunctionFlags.TypeDecorator)
-                .WithEvaluator(static (functionCall, decorated) => decorated switch
-                {
-                    DeclaredTypeExpression declaredType => declaredType with { Exported = functionCall },
-                    _ => decorated,
-                })
-                .WithValidator(static (decoratorName, decoratorSyntax, _, _, binder, _, diagnosticWriter) =>
-                {
-                    var decoratorTarget = binder.GetParent(decoratorSyntax);
-
-                    if (decoratorTarget is not ITopLevelNamedDeclarationSyntax)
+            if (featureProvider.CompileTimeImportsEnabled)
+            {
+                yield return new DecoratorBuilder(LanguageConstants.ExportPropertyName)
+                    .WithDescription("Allows a type or variable to be imported into other templates.")
+                    .WithFlags(FunctionFlags.TypeOrVariableDecorator)
+                    .WithEvaluator(static (functionCall, decorated) => decorated switch
                     {
-                        diagnosticWriter.Write(DiagnosticBuilder.ForPosition(decoratorSyntax).ExportDecoratorMustTargetStatement());
-                    }
-                })
-                .Build();
+                        DeclaredTypeExpression declaredType => declaredType with { Exported = functionCall },
+                        DeclaredVariableExpression declaredVariable => declaredVariable with { Exported = functionCall },
+                        _ => decorated,
+                    })
+                    .WithValidator(static (decoratorName, decoratorSyntax, _, _, binder, _, diagnosticWriter) =>
+                    {
+                        var decoratorTarget = binder.GetParent(decoratorSyntax);
+
+                        if (decoratorTarget is not ITopLevelNamedDeclarationSyntax)
+                        {
+                            diagnosticWriter.Write(DiagnosticBuilder.ForPosition(decoratorSyntax).ExportDecoratorMustTargetStatement());
+                        }
+
+                        if (decoratorTarget is not null && binder.GetSymbolInfo(decoratorTarget) is DeclaredSymbol targetedDeclaration)
+                        {
+                            var nonExportableSymbolsInClosure = SymbolicReferenceCollector.CollectSymbolsReferencedRecursive(binder, targetedDeclaration)
+                                .Where(s => s is not VariableSymbol &&
+                                    s is not TypeAliasSymbol &&
+                                    s is not ImportedSymbol &&
+                                    s is not WildcardImportSymbol &&
+                                    s is not LocalVariableSymbol)
```

Comment:
[nit] pattern matching makes this a little simpler:
`s is not (VariableSymbol or TypeAliasSymbol or ImportedSymbol or WildcardImportSymbol or LocalVariableSymbol)`